### PR TITLE
Allow ExecuteAsync cancellation propagation

### LINKED ISSUE #2811: BackgroundService should provide a linked token to ExecuteAsync
# Describe the bug

To allow cancellation occurring on start to flow through [ExecuteAsync](https://github.com/aspnet/Extensions/blob/master/src/Hosting/Abstractions/src/BackgroundService.cs#L33), ```BackgroundService``` should provide a linked token to ```ExecuteAsync```

### To Reproduce
1. Put ```await Task.Delay(Timeout.Infinite, stoppingToken);``` inside ```ExecuteAsync```
2. Do ```Ctrl+C```
3. App will wait indefinitely

### Expected behavior

* Cancel occurred

### Solution

```csharp
_stoppingCts  = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
_executingTask = ExecuteAsync(_stoppingCts.Token);
```

I can do a PR if needed

### PR DESCRIPTION
**Pull request by klettier**
Summary of the changes (Less than 80 chars)
 - Add test
 - Link the token provided to ```StartAsync``` to the one provided to the ```_executingTask ```

fixes #2811

### CODE REVIEW

File: src/Hosting/Abstractions/src/BackgroundService.cs [hunk 1]
```diff
@@ -29,6 +29,9 @@ public abstract class BackgroundService : IHostedService, IDisposable
         /// <param name="cancellationToken">Indicates that the start process has been aborted.</param>
         public virtual Task StartAsync(CancellationToken cancellationToken)
         {
+            // Create linked token
```
Comment (Tratcher) [inline src/Hosting/Abstractions/src/BackgroundService.cs]: This comment is redundant with the method name. How about explaining why?

File: src/Hosting/Abstractions/src/BackgroundService.cs [hunk 2]
```diff
@@ -13,7 +13,7 @@ namespace Microsoft.Extensions.Hosting
     public abstract class BackgroundService : IHostedService, IDisposable
     {
         private Task _executingTask;
-        private readonly CancellationTokenSource _stoppingCts = new CancellationTokenSource();
+        private CancellationTokenSource _stoppingCts;
```
Comment (Tratcher) [inline src/Hosting/Abstractions/src/BackgroundService.cs]: Dispose now has to handle the possibility that StartAsync was never called to initialize _stoppingCts. It should be updated to `_stoppingCts?.Cancel();`.

**Comment by sertunc** [comment]
Thanks a lot

**Comment by klettier** [comment]
Guys, do you when this will be the released ?

I have currently a case (**with current nuget 3.1.0**) that prevent me from cancelling the ```RunAsync``` without this change.

Bellow you can find an example of code that will hang forever.

```csharp
[Fact]
public async Task Cancel()
{
    var cancellationTokenSource = new CancellationTokenSource();

    await Host.CreateDefaultBuilder()
              .ConfigureServices(services =>
                {
                    services.AddHostedService(s => new CancellerService(cancellationTokenSource));
                })
              .Build()
              .RunAsync(cancellationTokenSource.Token);
}

class CancellerService : BackgroundService
{
    private readonly CancellationTokenSource tokenSource;
    System.Threading.Timer timer;

    public CancellerService(CancellationTokenSource tokenSource)
    {
        this.tokenSource = tokenSource;
        timer = new System.Threading.Timer(TimerCallback, null, TimeSpan.FromSeconds(5), TimeSpan.Zero);
    }

    void TimerCallback(object state)
    {
        tokenSource.Cancel();
    }

    protected override Task ExecuteAsync(CancellationToken stoppingToken)
    {
        var blocking = new BlockingCollection<int>();
        
        blocking.Take(stoppingToken);

        return Task.CompletedTask;
    }

    public override Task StopAsync(CancellationToken cancellationToken)
    {
        return base.StopAsync(cancellationToken);
    }
}
```

Thx

**Comment by Tratcher** [comment]
This will be included in the next major release (5.0) that has its first preview coming soon. The bar is very high for adding any patches to 3.1, it has to be a high impact issue with no workaround.

A workaround for the example above would be to add `await Task.Yield()` at the start of ExecuteAsync. That's a known issue we're addressing separately. https://github.com/dotnet/extensions/issues/2149

**Comment by klettier** [comment]
> This will be included in the next major release (5.0) that has its first preview coming soon. The bar is very high for adding any patches to 3.1, it has to be a high impact issue with no workaround.
> 
> A workaround for the example above would be to add `await Task.Yield()` at the start of ExecuteAsync. That's a known issue we're addressing separately. #2149

Thx a lot

**Pull request merged**